File: backend/app/models/mathlab_contest.py
"""
MathLab 课堂竞赛模型
"""

from datetime import datetime
from enum import Enum

from sqlalchemy import (
    Boolean,
    Column,
    DateTime,
    Enum as SQLEnum,
    Float,
    ForeignKey,
    Integer,
    JSON,
    String,
    UniqueConstraint,
    Index,
)
from sqlalchemy.orm import relationship

from app.core.database import Base


class MathlabContestStatus(str, Enum):
    PREPARING = "preparing"
    RUNNING = "running"
    ENDED = "ended"


class MathlabContest(Base):
    """一场 MathLab 课堂竞赛（挂在 ClassSession + SimCell）"""

    __tablename__ = "mathlab_contests"

    id = Column(Integer, primary_key=True, index=True)
    session_id = Column(
        Integer, ForeignKey("class_sessions.id", ondelete="CASCADE"), nullable=False, index=True
    )
    cell_id = Column(Integer, ForeignKey("cells.id", ondelete="SET NULL"), nullable=True, index=True)
    teacher_id = Column(Integer, ForeignKey("users.id"), nullable=False, index=True)
    task_id = Column(String(32), nullable=False)
    status = Column(
        SQLEnum(MathlabContestStatus),
        default=MathlabContestStatus.RUNNING,
        nullable=False,
        index=True,
    )
    time_limit_sec = Column(Integer, nullable=True)
    allow_resubmit = Column(Boolean, default=False, nullable=False)
    pass_threshold = Column(Integer, default=85, nullable=False)
    settings = Column(JSON, nullable=True, default=dict)
    started_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    ended_at = Column(DateTime, nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)

    session = relationship("ClassSession", foreign_keys=[session_id])
    submissions = relationship(
        "MathlabContestSubmission",
        back_populates="contest",
        cascade="all, delete-orphan",
    )

    def __repr__(self) -> str:
        return f"<MathlabContest(id={self.id}, session_id={self.session_id}, status={self.status})>"


class MathlabContestSubmission(Base):
    """学生竞赛提交"""

    __tablename__ = "mathlab_contest_submissions"
    __table_args__ = (
        UniqueConstraint("contest_id", "student_id", name="uq_mathlab_contest_student"),
        Index("ix_mathlab_contest_submissions_contest_score", "contest_id", "final_score"),
    )

    id = Column(Integer, primary_key=True, index=True)
    contest_id = Column(
        Integer,
        ForeignKey("mathlab_contests.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    student_id = Column(Integer, ForeignKey("users.id"), nullable=False, index=True)
    auto_score = Column(Float, nullable=False, default=0)
    auto_passed = Column(Boolean, default=False, nullable=False)
    final_score = Column(Float, nullable=False, default=0)
    passed = Column(Boolean, default=False, nullable=False)
    elapsed_sec = Column(Float, nullable=True)
    payload = Column(JSON, nullable=True, default=dict)
    submitted_at = Column(DateTime, default=datetime.utcnow, nullable=False)

    contest = relationship("MathlabContest", back_populates="submissions")
    student = relationship("User", foreign_keys=[student_id])

    def __repr__(self) -> str:
        return f"<MathlabContestSubmission(contest_id={self.contest_id}, student_id={self.student_id})>"

File: frontend/src/services/mathlabContest.ts
/**
 * MathLab 课堂竞赛 API
 */

import api from './api'
import type {
  MathlabContest,
  MathlabContestLeaderboard,
  MathlabContestStartRequest,
  MathlabContestSubmitRequest,
  MathlabContestSubmission,
} from '../types/mathlabContest'

function mapContest(raw: Record<string, unknown>): MathlabContest {
  return {
    id: Number(raw.id),
    sessionId: Number(raw.session_id ?? raw.sessionId),
    cellId: raw.cell_id != null ? Number(raw.cell_id) : raw.cellId != null ? Number(raw.cellId) : undefined,
    teacherId: Number(raw.teacher_id ?? raw.teacherId),
    taskId: String(raw.task_id ?? raw.taskId),
    status: String(raw.status).toLowerCase() as MathlabContest['status'],
    timeLimitSec: raw.time_limit_sec != null ? Number(raw.time_limit_sec) : raw.timeLimitSec != null ? Number(raw.timeLimitSec) : undefined,
    allowResubmit: Boolean(raw.allow_resubmit ?? raw.allowResubmit),
    passThreshold: Number(raw.pass_threshold ?? raw.passThreshold ?? 85),
    settings: (raw.settings as Record<string, unknown>) || {},
    startedAt: String(raw.started_at ?? raw.startedAt),
    endedAt: raw.ended_at ? String(raw.ended_at) : raw.endedAt ? String(raw.endedAt) : undefined,
    endsAt: raw.ends_at ? String(raw.ends_at) : raw.endsAt ? String(raw.endsAt) : undefined,
  }
}

function mapSubmission(raw: Record<string, unknown>): MathlabContestSubmission {
  return {
    id: Number(raw.id),
    contestId: Number(raw.contest_id ?? raw.contestId),
    studentId: Number(raw.student_id ?? raw.studentId),
    studentName: raw.student_name != null ? String(raw.student_name) : raw.studentName != null ? String(raw.studentName) : undefined,
    autoScore: Number(raw.auto_score ?? raw.autoScore),
    autoPassed: Boolean(raw.auto_passed ?? raw.autoPassed),
    finalScore: Number(raw.final_score ?? raw.finalScore),
    passed: Boolean(raw.passed),
    elapsedSec: raw.elapsed_sec != null ? Number(raw.elapsed_sec) : raw.elapsedSec != null ? Number(raw.elapsedSec) : undefined,
    payload: (raw.payload as Record<string, unknown>) || {},
    submittedAt: String(raw.submitted_at ?? raw.submittedAt),
    rank: raw.rank != null ? Number(raw.rank) : undefined,
  }
}

export const mathlabContestService = {
  async start(sessionId: number, body: MathlabContestStartRequest): Promise<MathlabContest> {
    const res = await api.post(
      `/classroom-sessions/sessions/${sessionId}/mathlab-contest/start`,
      {
        cell_id: body.cellId,
        task_id: body.taskId,
        time_limit_sec: body.timeLimitSec,
        allow_resubmit: body.allowResubmit ?? false,
        pass_threshold: body.passThreshold ?? 85,
        settings: body.settings,
      }
    )
    return mapContest(res as Record<string, unknown>)
  },

  async getActive(sessionId: number): Promise<MathlabContest | null> {
    const res = await api.get(
      `/classroom-sessions/sessions/${sessionId}/mathlab-contest/active`
    )
    if (!res) return null
    return mapContest(res as Record<string, unknown>)
  },

  async end(contestId: number): Promise<MathlabContest> {
    const res = await api.post(`/classroom-sessions/mathlab-contest/${contestId}/end`)
    return mapContest(res as Record<string, unknown>)
  },

  async updateTask(contestId: number, taskId: string): Promise<MathlabContest> {
    const res = await api.patch(`/classroom-sessions/mathlab-contest/${contestId}/task`, {
      task_id: taskId,
    })
    return mapContest(res as Record<string, unknown>)
  },

  async getLeaderboard(contestId: number): Promise<MathlabContestLeaderboard> {
    const res = await api.get(`/classroom-sessions/mathlab-contest/${contestId}/leaderboard`)
    const r = res as Record<string, unknown>
    return {
      contest: mapContest((r.contest as Record<string, unknown>) || r),
      submissions: ((r.submissions as Record<string, unknown>[]) || []).map(mapSubmission),
      submittedCount: Number(r.submitted_count ?? r.submittedCount ?? 0),
      totalStudents: Number(r.total_students ?? r.totalStudents ?? 0),
    }
  },

  async submit(contestId: number, body: MathlabContestSubmitRequest): Promise<MathlabContestSubmission> {
    const res = await api.post(`/classroom-sessions/mathlab-contest/${contestId}/submit`, {
      auto_score: body.autoScore,
      auto_passed: body.autoPassed,
      elapsed_sec: body.elapsedSec,
      payload: body.payload,
    })
    return mapSubmission(res as Record<string, unknown>)
  },

  async updateScore(
    submissionId: number,
    finalScore: number,
    passed?: boolean
  ): Promise<MathlabContestSubmission> {
    const res = await api.patch(
      `/classroom-sessions/mathlab-contest/submissions/${submissionId}/score`,
      { final_score: finalScore, passed }
    )
    return mapSubmission(res as Record<string, unknown>)
  },
}
